roboteam_ai/src/utilities/World.cpp

```diff
@@ -131,28 +134,95 @@ std::shared_ptr<roboteam_msgs::WorldBall> World::getBall() {
+    Vector2 dribbleRight = robotPos
+            + Vector2(Constants::ROBOT_RADIUS(), 0).rotate(robotOrientation + Constants::DRIBBLER_ANGLE_OFFSET());
+    // if the ball is very close (or on top of the robot?)
     if (control::ControlUtils::pointInTriangle(ballPos, robotPos, dribbleLeft, dribbleRight)) {
-        return true;
+        return 0.0;
     }
-        // else check the rectangle in front of the robot.
-    else
-        return control::ControlUtils::pointInRectangle(ballPos, dribbleLeft, dribbleRight,
-                dribbleRight + Vector2(frontDist, 0).rotate(robotOrientation),
-                dribbleLeft + Vector2(frontDist, 0).rotate(robotOrientation));
-}
+    else if (control::ControlUtils::pointInRectangle(ballPos, dribbleLeft, dribbleRight,
+            dribbleRight + Vector2(Constants::MAX_BALL_BOUNCE_RANGE(), 0).rotate(robotOrientation),
+            dribbleLeft + Vector2(Constants::MAX_BALL_BOUNCE_RANGE(), 0).rotate(robotOrientation))) {
+        return control::ControlUtils::distanceToLine(ballPos, dribbleLeft, dribbleRight);
+    }
+    // default return
+    return - 1.0;
 
+}
+void World::updateBallPossession(roboteam_msgs::World &_world) {
+    OurBotsBall.clear();
+    TheirBotsBall.clear();
+    // calculate if a bot is in dribbling range and if so what range it is at.
+    for (auto ourBot :_world.us) {
+        double dist = findBallDist(ourBot, _world.ball);
+        if (dist != - 1.0) {
+            OurBotsBall[ourBot.id] = dist;
+        }
+    }
+    for (auto theirBot: _world.them) {
+        double dist = findBallDist(theirBot, _world.ball);
+        if (dist != - 1.0) {
+            TheirBotsBall[theirBot.id] = dist;
+        }
+    }
+}
+// uses MAX_BALL_RANGE as default max Dist
+bool World::BotHasBall(int id, bool ourTeam, double maxDistToBall) {
+    if (ourTeam){
+        return ourBotHasBall(id,maxDistToBall);
+    }
+    else{
+        return theirBotHasBall(id,maxDistToBall);
+    }
+}
+bool World::ourBotHasBall(int id, double maxDistToBall){
+    std::lock_guard<std::mutex> lock(worldMutex);
+    if (OurBotsBall.find(id)!=OurBotsBall.end()){
+        if (OurBotsBall[id]<=maxDistToBall){
+            return true;
+        }
+    }
+    return false;
+}
+bool World::theirBotHasBall(int id, double maxDistToBall){
+    std::lock_guard<std::mutex> lock(worldMutex);
+    if (TheirBotsBall.find(id)!=TheirBotsBall.end()){
+        if (TheirBotsBall[id]<=maxDistToBall){
+            return true;
+        }
+    }
+    return false;
+}
+// picks the bot that has the ball and is closest to it.
+int World::whichBotHasBall(bool ourTeam) {
+    std::lock_guard<std::mutex> lock(worldMutex);
+    double maxDist = 100;
+    int bestId = - 1;
+    if (ourTeam) {
+        for (auto bot: OurBotsBall) {
+            if (bot.second < maxDist) {
+                maxDist = bot.second;
+                bestId = bot.first;
+            }
```

Comment:
this whole file feels like it has a lot of duplicates and it can be simplified quite a bit with different abstractions